exchange-proxy/features/meta-transactions.md

```diff
@@ -0,0 +1,212 @@
+        external
+        view
+        returns (bool);
+
+    // Get the hash of a meta-transaction.
+    function getMetaTransactionHash(MetaTransactionData mtx)
+        external
+        view
+        returns (bytes32);
+}
+```
+
+### Implementation
+
+All external execution functions just redirect to a private `_executeMetaTransactionPrivate()` function.
+
+```
+function _executeMetaTransaction(
+    address sender,
+    MetaTransactionData mtx,
+    bytes signature
+)
+    public
+    payable
+    returns (bytes returnData)
+{
+    returnData = _executeMetaMetaTransactionPrivate(sender, mtx, signature);
+}
+
+function executeMetaTransaction(MetaTransactionData mtx, bytes signature)
+    external
+    payable
+    returns (bytes returnData)
+{
+    returnData = _executeMetaTransactionPrivate(msg.sender, mtx, signature);
+}
+
+function executeMetaTransactions(MetaTransactionData[] mtxs, bytes[] signatures)
+    external
+    payable
+    returns (bytes[] returnDatas)
+{
+    returnDatas = new bytes[](mtxs.length);
+    for (uint256 i = 0; i < mtxs.length; ++i) {
+        returnDatas[i] = _executeMetaTransactionPrivate(
+            msg.sender,
+            mtxs[i],
+            signatures[i]
+        );
+    }
+}
+
+function _executeMetaTransactionPrivate(
+    address sender,
+    MetaTransactionData mtx,
+    bytes signature
+)
+    private
+    returns (bytes returnData)
+{
+    LibMetaTransactionsStorage.Storage stor = LibMetaTransactionsStorage.getStorage();
+    bytes32 hash = getTransactionHash(mtx);
+
+    require(mtx.sender == address(0) || mtx.sender == sender);
+    require(!stor.isTransactionExecuted(hash));
+    require(
+        ISignatureValidator(address(this)).isValidSignature(
+            abi.encode(mtx),
+            hash,
+            mtx.signer,
+            signature
+    ));
+    stor.isTransactionExecuted[hash] = true;
+
+    bool success;
+    bytes4 selector = mtx.callData.readBytes4();
+    if (selector == ITransformERC20.transformERC20.selector) {
+        // Decode call args.
+        (
+            IERC20Token inputToken,
+            IERC20Token outputToken,
+            uint256 inputTokenAmount,
+            uint256 minOutputTokenAmount,
+            Transformation[] transformations
+        ) = abi.decode(
+            mtx.callData,
+            (IERC20Token, IERC20Token, uint256, uint256, Transformation[])
+        );
+        // Call the internal `_transformERC20()`.
+        (success, returnData) = address(this).call{value: mtx.value}(
+            abi.encodeWithSelector(
+            ITransformERC20._transformERC20
+            keccak256(mtx.callData),
+            mtx.signer,
+            inputToken,
+            outputToken,
+            inputTokenAmount,
+            minOutputTokenAmount,
+            transformations
+        ));
+    } else {
+        revert('UNSUPPORTED_TRANSACTION');
+    }
+    if (!success) {
+        returnData.rrevert();
+    }
+}
+```
+
```

Comment:
I think it would be good to add a section for our integrators where we emphasize that (i) this provides mtx's for EP transformers (+ link to spec on transformers), (ii) does not support Exchange V3 transactions, and (iii) works strictly with `eth_sign` and `EIP712`.